transport-native-kqueue/src/main/java/io/netty/channel/kqueue/BsdSocket.java

```diff
@@ -80,6 +94,86 @@ long sendFile(DefaultFileRegion src, long baseOffset, long offset, long length)
+            throws IOException {
+        checkNotNull(destination, "Destination InetSocketAddress cannot be null.");
+
+        boolean sourceIPv6;
+        byte[] sourceAddress;
+        int sourceScopeId;
+        int sourcePort;
+        if (source == null) {
+            sourceIPv6 = false;
+            sourceAddress = null;
+            sourceScopeId = 0;
+            sourcePort = 0;
+        } else {
+            InetAddress sourceInetAddress = source.getAddress();
+            sourceIPv6 = sourceInetAddress instanceof Inet6Address;
+            if (sourceIPv6) {
+                sourceAddress = sourceInetAddress.getAddress();
+                sourceScopeId = ((Inet6Address) sourceInetAddress).getScopeId();
+            } else {
+                // convert to ipv4 mapped ipv6 address;
+                sourceScopeId = 0;
+                sourceAddress = ipv4MappedIpv6Address(sourceInetAddress.getAddress());
+            }
+            sourcePort = source.getPort();
+        }
+
+        InetAddress destinationInetAddress = destination.getAddress();
+        boolean destinationIPv6 = destinationInetAddress instanceof Inet6Address;
+        byte[] destinationAddress;
+        int destinationScopeId;
+        if (destinationIPv6) {
+            destinationAddress = destinationInetAddress.getAddress();
+            destinationScopeId = ((Inet6Address) destinationInetAddress).getScopeId();
+        } else {
+            // convert to ipv4 mapped ipv6 address;
+            destinationScopeId = 0;
+            destinationAddress = ipv4MappedIpv6Address(destinationInetAddress.getAddress());
+        }
+        int destinationPort = destination.getPort();
+
+        long iovAddress;
+        int iovCount;
+        int iovDataLength;
+        if (data == null || data.count() == 0) {
+            iovAddress = 0;
+            iovCount = 0;
+            iovDataLength = 0;
+        } else {
+            iovAddress = data.memoryAddress(0);
+            iovCount = data.count();
+            long size = data.size();
+            if (size > Integer.MAX_VALUE) {
+                throw new IOException("IovArray.size() too big: " + size + " bytes.");
+            }
+            iovDataLength = (int) size;
+        }
+
+        int result = connectx(intValue(),
+                UNSPECIFIED_SOURCE_INTERFACE, sourceIPv6, sourceAddress, sourceScopeId, sourcePort,
+                destinationIPv6, destinationAddress, destinationScopeId, destinationPort,
+                flags, iovAddress, iovCount, iovDataLength);
+        if (result == ERRNO_EINPROGRESS_NEGATIVE) {
+            // This is normal for non-blocking sockets.
+            // We'll know the connection has been established when the socket is selectable for writing.
+            // Tell the channel the data was written, so the outbound buffer can update its position.
+            return iovDataLength;
```

Comment:
Is this correct to assume that all data will be written when the connect is in progress? I see this on the man page:

> In general, the communication domain makes the final decision on the
     amount of data that may get transmitted at connection establishment time.

As a reference in [`Socket`](https://github.com/netty/netty/blob/4.1/transport-native-unix-common/src/main/java/io/netty/channel/unix/Socket.java#L137-L141) when we get this error, we assume nothing was written. Although it is a different method, so not sure if these two are equatable.

Reply:
The `connectx(2)` behaves differently and will copy the data to the kernel-side socket buffer regardless of whether a TFO connection was established or not.